site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java

```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
-    			group.getProperties().addProperty(group.GROUP_PROP_WSETUP_CREATED, Boolean.TRUE.toString());
+        int groupCount = 0;
+        List<Group> groupList = new ArrayList<>();
+        while( groupCount < numOfGroups )
+        {
+            // Create the group
+            Group group = site.addGroup();
+            group.getProperties().addProperty( Group.GROUP_PROP_WSETUP_CREATED, Boolean.TRUE.toString() );
+            group.setTitle( groupTitle + "-" + (groupCount + 1) );
 
-    			//Title
-    			StringBuffer title = new StringBuffer();
+            int userCount = 0;
+            while( userCount < numOfUsersPerGroup && userIDs.size() > 0 )
+            {
+                // Pick a random user
+                int index = (int)(Math.random() * (userIDs.size() - 1));
+                String userID = userIDs.get( index );
+                Member member = site.getMember( userID );
 
-    			title.append(groupTitle);
+                // Add the user to the group
+                group.addMember( userID, member.getRole().getId(), member.isActive(), false );
+                userIDs.remove( index );
+                userCount++;
+            }
 
-    			title.append("-");
-    			title.append(groupCount+1);
-    			group.setTitle(title.toString());
+            groupCount++;
+            groupList.add( group );
+        }
 
-    			int userCount = 0;
+        // All the groups have been created, but there are some users still left to be assigned (remainders)
+        while( userIDs.size() > 0 )
+        {
+            // Pick a random user
+            int userIndex = (int)(Math.random() * (userIDs.size() - 1));
+            String userID = userIDs.get( userIndex );
 
-    			while(userCount < numOfUsersPerGroup && usersList.size() > 0){
-    				int index = (int)(Math.random() * (usersList.size() - 1));
-    				String userId = usersList.get(index);
-    				Member member = site.getMember(userId);
-    				group.addMember(userId, member.getRole().getId(), member.isActive(), false);
+            // Pick a random group
+            int groupIndex = (int)(Math.random() * (groupList.size() - 1));
+            Group group = groupList.get( groupIndex );
 
-    				//remove this user now:
-    				usersList.remove(index);
+            // Add the user to the group
+            Member member = site.getMember( userID );
+            group.addMember( userID, member.getRole().getId(), member.isActive(), false );
+            userIDs.remove( userIndex );
+        }
+    }
 
-    				userCount++;
-    			}       		
+    /**
+     * Create random groups of users from the given role.
+     * This could be random number of users in specified number of groups,
+     * or vice versa.
+     * 
+     * @param roleList a list of roles (should only ever be one in the list)
+     * @param unit the number of groups or users per group requested
+     */
+    private void createRandomGroupsForRole( List<String> roleList, int unit )
```

Comment:
if there should only ever be one role, change the parameter to take a single role instead of a list